File: system-test/java/src/main/java/com/optivem/shop/testkit/core/usecase/shop/usecases/BrowseCoupons.java
package com.optivem.shop.testkit.core.usecase.shop.usecases;

import com.optivem.shop.testkit.core.shared.UseCaseContext;
import com.optivem.shop.testkit.core.shared.UseCaseResult;
import com.optivem.shop.testkit.core.usecase.shop.usecases.base.BaseShopUseCase;
import com.optivem.shop.testkit.driver.port.shop.ShopDriver;
import com.optivem.shop.testkit.driver.port.shop.dtos.BrowseCouponsResponse;

public class BrowseCoupons extends BaseShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification> {
    public BrowseCoupons(ShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    @Override
    public UseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification> execute() {
        var result = driver.browseCoupons();
        return new UseCaseResult<>(result, context, BrowseCouponsVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/shop/usecases/BrowseCoupons.ts
import type { ShopDriver } from '../../../../../driver/port/shop/shop-driver.js';
import type { BrowseCouponsResponse } from '../../../../../driver/port/shop/dtos/BrowseCouponsResponse.js';
import { UseCaseResult } from '../../../shared/use-case-result.js';
import type { UseCaseContext } from '../../../shared/use-case-context.js';
import { BaseShopUseCase } from './base/BaseShopUseCase.js';
import { BrowseCouponsVerification } from './BrowseCouponsVerification.js';

export class BrowseCoupons extends BaseShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification> {
  constructor(driver: ShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  async execute(): Promise<UseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification>> {
    const result = await this.driver.browseCoupons();

    return new UseCaseResult(
      result,
      this.context,
      (response, ctx) => new BrowseCouponsVerification(response, ctx),
    );
  }
}
